### ISSUE TITLE
The model after INT8 quantization is slower than before and with higher RAM consumption

### ISSUE THREAD

**Issue by jsvir** [NVIDIA/TensorRT#3018]
## Description

Hi, I try to use pytorch-quantization on MarbleNet model which is trained and saved in FP32.
When I apply quantization and use pytorch profiler to measure the RAM and CPU speed I get better results before quantization.
I check only encoder part which is ResNet-like module with Conv1d layers.

In addition, when I export the model to onnx I also get a larger model in KBs than the original one which exported with the same parameters. 

## Environment

**pytorch-quantization**:  2.1.2 

**NVIDIA GPU**: GTX 1080

**NVIDIA Driver Version**:  465.19.01

**CUDA Version**: 11.3

Operating System: Linux 

Python Version : 3.8.10

PyTorch Version (if applicable):

Container (if so, version):  nvcr.io/nvidia/pytorch :21.10-py3

ONNX: 1.8

**Model link**:

https://catalog.ngc.nvidia.com/orgs/nvidia/teams/nemo/models/vad_telephony_marblenet

**My code**:

```
import torch
from torch.profiler import profile, ProfilerActivity
import numpy as np
from omegaconf import OmegaConf
from nemo.collections.asr.models import EncDecClassificationModel
from pytorch_quantization import quant_modules


if __name__ == "__main__":
    quant_modules.initialize()
    args = OmegaConf.create(dict(
        vad_model="vad_telephony_marblenet.nemo",
    ))
    vad_model = EncDecClassificationModel.restore_from(restore_path=args.vad_model)
    vad_model.encoder.quantize = True
    print(vad_model.encoder)
    vad_model.preprocessor.featurizer.dither = 0.0
    vad_model.preprocessor.featurizer.pad_to = 0
    vad_model.eval()
    vad_model.to("cpu")
    random_wave = torch.tensor(np.random.randn(16000)).reshape(1, -1).float()
    wave_length = torch.tensor([16000]).reshape(1).long()
    random_spec, random_length = vad_model.preprocessor(input_signal=random_wave, length=wave_length)
    with profile(activities=[ProfilerActivity.CPU],
                 profile_memory=True, record_shapes=True) as prof:
        output = vad_model.encoder(audio_signal=random_spec, length=random_length)
    print(prof.key_averages().table(sort_by="self_cpu_memory_usage", row_limit=10))
    print("output.size()=", output[0].size())
```

**Output before quantization**:
```
[W CPUAllocator.cpp:219] Memory block of unknown size was allocated before the profiling started, profiler results will not include the deallocation event
--------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                            Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg       CPU Mem  Self CPU Mem    # of Calls
--------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                     aten::empty         3.66%     371.000us         3.66%     371.000us       3.312us     613.50 Kb     613.50 Kb           112
                       aten::mul         7.24%     733.000us        11.16%       1.130ms      56.500us     606.00 Kb     598.11 Kb            20
                   aten::resize_         2.30%     233.000us         2.30%     233.000us       7.281us     394.53 Kb     394.53 Kb            32
                       aten::add         0.81%      82.000us         0.81%      82.000us      27.333us      75.75 Kb      75.75 Kb             3
             aten::empty_strided         1.39%     141.000us         1.39%     141.000us       7.050us       7.89 Kb       7.89 Kb            20
                        aten::lt        10.45%       1.058ms        10.45%       1.058ms      17.633us       2.01 Kb       2.01 Kb            60
          _masked_conv_init_lens         5.42%     549.000us        13.90%       1.407ms      70.350us      15.94 Kb         149 b            20
                aten::is_nonzero         0.44%      45.000us         1.02%     103.000us       5.150us           0 b           0 b            20
                        defaults         0.07%       7.000us         0.07%       7.000us       1.167us           0 b           0 b             6
                    aten::arange         3.19%     323.000us         7.09%     718.000us      17.950us      31.56 Kb           0 b            40
--------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 10.123ms

output.size()= torch.Size([1, 128, 101])
```
**Output after quantization**:

```
[W CPUAllocator.cpp:219] Memory block of unknown size was allocated before the profiling started, profiler results will not include the deallocation event
--------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                            Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg       CPU Mem  Self CPU Mem    # of Calls
--------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                       aten::mul         5.21%       1.666ms         6.54%       2.091ms      33.190us       1.59 Mb       1.58 Mb            63
                       aten::div         4.56%       1.456ms         4.56%       1.456ms      16.930us       1.00 Mb       1.00 Mb            86
                     aten::clamp         2.81%     898.000us         2.81%     898.000us      20.884us    1021.50 Kb    1021.50 Kb            43
                     aten::empty         2.99%     957.000us         2.99%     957.000us       4.833us     613.84 Kb     613.84 Kb           198
                   aten::resize_         0.72%     231.000us         0.72%     231.000us       7.219us     394.53 Kb     394.53 Kb            32
                       aten::add         0.28%      88.000us         0.28%      88.000us      29.333us      75.75 Kb      75.75 Kb             3
             aten::empty_strided         3.55%       1.134ms         3.55%       1.134ms       5.906us       8.56 Kb       8.56 Kb           192
                        aten::lt         6.94%       2.219ms         9.33%       2.981ms      28.942us       2.05 Kb       1.89 Kb           103
                       aten::neg         1.29%     411.000us         1.29%     411.000us       9.558us         172 b         172 b            43
          _masked_conv_init_lens         1.25%     401.000us         4.15%       1.325ms      66.250us      15.94 Kb         145 b            20
--------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 31.960ms

output.size()= torch.Size([1, 128, 101])
```

To calibrate the model I follow the example [here](https://github.com/NVIDIA/TensorRT/blob/master/tools/pytorch-quantization/examples/calibrate_quant_resnet50.ipynb) and I apply max calibration.

Thanks in advance,

Jonathan

**Comment by zerollzeng** [comment]
> When I apply quantization and use pytorch profiler to measure the RAM and CPU speed I get better results before quantization.

If you mean INT8 accuracy is slightly worse than FP32, then I think it's expected.



> In addition, when I export the model to onnx I also get a larger model in KBs than the original one which exported with the same parameters.

It's also expected since the new onnx add extra Q/DQ nodes.

**Comment by jsvir** [comment]
Hi @zerollzeng,
I do not mean accuracy, but inference speed and memory consumption. Btw, all tests were done with CPU only.

Thanks

**Comment by zerollzeng** [comment]
> Btw, all tests were done with CPU only.

You should test the quantized model with GPU and compare the perf with FP32 inference?

**Comment by ttyio** [comment]
closing since no activity for more than 3 weeks, pls reopen if you still have question, thanks!
